File: benchmarks-QuantumOptics.jl/displace.jl
using QuantumOptics
using BenchmarkTools
include("benchmarkutils.jl")

name = "displace"

samples = 3
evals = 10
cutoffs = [10:10:150;]

function setup(N)
    b = FockBasis(N-1)
    alpha = log(N)
    b, alpha
end

function f(b, alpha)
    displace(b, alpha)
end

println("Benchmarking: ", name)
print("Cutoff: ")
checks = Dict{Int, Float64}()
results = []
for N in cutoffs
    print(N, " ")
    b, alpha = setup(N)
    checks[N] = expect(destroy(b), f(b, alpha))
    t = @belapsed f($b, $alpha) samples=samples evals=evals
    push!(results, Dict("N"=>N, "t"=>t))
end
println()

benchmarkutils.check(name, checks)
benchmarkutils.save(name, results)

File: benchmarks-QuTiP/displace.py
import qutip as qt
import numpy as np
import benchmarkutils

name = "displace"

samples = 3
evals = 10
cutoffs = range(10, 151, 10)

def setup(N):
    alpha = np.log(N)
    return alpha

def f(N, alpha):
    return qt.displace(N, alpha)

print("Benchmarking:", name)
print("Cutoff: ", end="", flush=True)
checks = {}
results = []
for N in cutoffs:
    print(N, "", end="", flush=True)
    alpha = setup(N)
    checks[N] = qt.expect(qt.destroy(N), f(N, alpha))
    t = benchmarkutils.run_benchmark(f, N, alpha, samples=samples, evals=evals)
    results.append({"N": N, "t": t})
print()

benchmarkutils.check(name, checks, 0.05)
benchmarkutils.save(name, results)
